go-controller/pkg/clustermanager/userdefinednetwork/controller.go

```diff
@@ -135,6 +136,42 @@ func (c *Controller) syncUserDefinedNetwork(udn *userdefinednetworkv1.UserDefine
 		return nil, nil
 	}
 
+	if !udn.DeletionTimestamp.IsZero() { // udn is being  deleted
+		if controllerutil.ContainsFinalizer(udn, template.FinalizerUserDefinedNetwork) {
+
+			if nad != nil &&
+				metav1.IsControlledBy(nad, udn) &&
+				controllerutil.ContainsFinalizer(nad, template.FinalizerUserDefinedNetwork) {
+
+				// TODO: verify no pod is using the NAD before removing finalizer from NAD
+
+				controllerutil.RemoveFinalizer(nad, template.FinalizerUserDefinedNetwork)
```

Comment:
When UDN is created, the controller add finalizer to the UDN object and create the NAD with a finalizer.
 
When the UDN is deleted, the controller first remove the finalizer from the NAD, when succeed it will remove the finalizer from the UDN and allow it to dispose (by the gurbage collector).
The next commit adds a check if the NAD is being used (pod connected to it), and remove the finalizer from the NAD accordingly.

Finalizer is placed on the NAD to prevent it from being deleted manually or before the UDN is deleted, or when being used (pod connected to it).

Reply:
makes sense, thanks.